include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h

```diff
@@ -1328,12 +1328,12 @@ __pattern_stable_partition(__hetero_tag<_BackendTag> __tag, _ExecutionPolicy&& _
 
```

Comment:
Do we have any guarantee that the first `__pattern_walk2()` call above returns?
I think these are independent from each other, and were both depending on the destructors for synchronization (or so we thought...).

I think it may need to be made synchronous as well.  It would be nice to have it return an event which could be waited on so that these functions could run concurrently, but that is likely beyond the scope of this PR.

Reply:
1) These patterns (pattern_merge and pattern_walk2) have  an implicit dependence via SYCL dependency graph.
2) In other pattern (__pattern_stable_partition) we have dependent working patterns, and those can be launched asynchrony and we have to wait two events by `wait(e1, e2)`. There is a comment about that in the code.

Reply:
For `__pattern_stable_partition`, line 1325, we call `__pattern_walk2` without synchronization.  It has no dependencies with the blocking call on line 1331 as far as I can see.  

Therefore, we have no guarantee that the call from line 1325 completes before we return from `__pattern_stable_partition`.  Ideally `__pattern_walk2` would return a future or event we could wait on, but it does not, so I think we just need to call it in a blocking fashion, unless we want more refactoring.

I'm missing where you mention we wait on 2 events as you describe: `wait(e1, e2)`.

Reply:
yes, I agree with that.